framework/src/modules/chain/bft/bft.js

```diff
@@ -14,273 +14,135 @@
+		const bftMigrationHeight =
+			this.constants.startingHeight - this.constants.activeDelegates * 2;
+
+		// Choose max between stored finalized height or migration height
+		const finalizedHeight = Math.max(finalizedHeightStored, bftMigrationHeight);
+
+		// Initialize consensus manager
+		return new FinalityManager({
+			finalizedHeight,
+			activeDelegates: this.constants.activeDelegates,
+		});
 	}
 
 	/**
-	 * Use existing block headers and re-compute all information
+	 * Get the last block height from storage
+	 *
+	 * @return {Promise<number>}
+	 * @private
 	 */
-	recompute() {
-		this.state = {};
-		this.finalizedHeight = 0;
-		this.prevotedConfirmedHeight = 0;
-		this.preVotes = {};
-		this.preCommits = {};
-
-		this.headers.items.forEach(header => {
-			this.updatePreVotesPreCommits(header);
-			this.updatePreVotedAndFinalizedHeight();
-		});
+	async _getLastBlockHeight() {
+		const lastBlock = await this.BlockEntity.get(
+			{},
+			{ limit: 1, sort: 'height:desc' }
+		);
+		return lastBlock.length ? lastBlock[0].height : 0;
 	}
 
 	/**
-	 * Verify if the block header is good for current chain
+	 * Load blocks into consensus manager fetching from storage
 	 *
-	 * @param {BlockHeader} blockHeader
+	 * @param {integer} fromHeight - The start height to fetch and load
+	 * @param {integer} tillHeight - The end height to fetch and load
+	 * @return {Promise<void>}
 	 */
-	verifyBlockHeaders(blockHeader) {
-		// We need minimum processingThreshold to decide
-		// if prevotedConfirmedUptoHeight is correct
-		if (
-			this.headers.length >= this.processingThreshold &&
-			blockHeader.prevotedConfirmedUptoHeight !== this.prevotedConfirmedHeight
-		) {
-			throw new Error('Wrong prevotedConfirmedHeight in blockHeader.');
-		}
-
-		// Find top most block forged by same delegate
-		const delegateLastBlock = this.headers
-			.top(this.processingThreshold)
-			.reverse()
-			.find(
-				header => header.delegatePublicKey === blockHeader.delegatePublicKey
-			);
-
-		if (!delegateLastBlock) {
-			return true;
-		}
-
-		if (
-			delegateLastBlock.maxHeightPreviouslyForged ===
-				blockHeader.maxHeightPreviouslyForged &&
-			delegateLastBlock.height >= blockHeader.height
-		) {
-			// Violation of the fork choice rule as delegate moved to different chain
-			// without strictly larger maxHeightPreviouslyForged or larger height as
-			// justification. This in particular happens, if a delegate is double forging.
-			throw new Error(
-				'Violation of fork choice rule, delegate moved to a different chain'
-			);
-		}
+	async loadBlocks({ fromHeight, tillHeight }) {
+		const rows = await this.BlockEntity.get(
+			{ height_gte: fromHeight, height_lte: tillHeight },
+			{ limit: null, sort: 'height:asc' }
+		);
 
-		if (delegateLastBlock.height > blockHeader.maxHeightPreviouslyForged) {
-			throw new Error('Violates disjointness condition');
-		}
+		rows.forEach(row => {
+			if (row.version !== '2') return;
```

Comment:
For me, it makes more sense to throw an error here.

Reply:
Its the edge case for migration. When we migrate on a particular height, we assume that height  - 202 as finalized height. While loading we have to load all blocks higher from finalized height. In that case we can encounter blocks with older version and we need to just ignore those blocks to make BFT ready for further processing.